### PR TITLE
Initial bootstrap implementation.

### PR DESCRIPTION
**Pull request by vldm**
Closes: #319 
Closes: #320 

- Implemented basic network bootstrap.
- ConsensusMessages in one epoch from ALL future height now are silently added into queue.

### CODE REVIEW

File: node/src/chain_loader.rs [hunk 1]
```diff
@@ -0,0 +1,228 @@
+//
```
Comment (rtsisyk) [inline node/src/chain_loader.rs]: m.b. just `loader.rs` or even better `replication.rs`?

File: node/src/chain_loader.rs [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+use stegos_blockchain::Block;
+use stegos_crypto::hash::{Hash, Hashable, Hasher};
+use stegos_crypto::pbc::secure;
+use stegos_serialization::traits::ProtoConvert;
+
+use crate::consensus::SealedBlockMessage;
+use crate::NodeService;
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct BlocksRange {
```
Comment (rtsisyk) [inline node/src/chain_loader.rs]: BlockRange (without `s`)

File: node/src/lib.rs [hunk 3]
```diff
@@ -1338,9 +1360,14 @@ impl Future for NodeService {
...
                         NodeMessage::Transaction(msg) => self.handle_transaction(msg),
                         NodeMessage::Consensus(msg) => self.handle_consensus_message(msg),
-                        NodeMessage::SealedBlock(msg) => self.handle_sealed_block(msg),
+                        NodeMessage::SealedBlock(msg) => SealedBlockMessage::from_buffer(&msg)
+                            .and_then(|msg| self.handle_sealed_block(msg)),
                         NodeMessage::ConsensusTimer(_now) => self.handle_consensus_timer(),
                         NodeMessage::VRFMessage(msg) => self.handle_vrf_message(msg),
+                        NodeMessage::ChainLoaderMessage(msg) => {
+                            chain_loader::ChainLoaderMessage::from_buffer(&msg.data)
+                                .and_then(|data| self.handle_chain_loader_message(msg.from, data))
```
Comment (rtsisyk) [inline node/src/lib.rs]: Please make a patch on the top of `dev` to change all similar places to use this approach.

File: node/src/chain_loader.rs (hunk 2 above)
Comment (rtsisyk) [reply node/src/chain_loader.rs]: BTW, the only use of this structure is `RequestBlocks`. Does it make sense to create a separate structure? Let's integrate it into `RequestBlocks` to simplify things.

File: node/src/chain_loader.rs [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    fn hash(&self, state: &mut Hasher) {
+        self.range.hash(state);
+    }
+}
+
+impl RequestBlocks {
+    pub fn new(current_hash: Hash, block_hash: Hash) -> RequestBlocks {
+        let range = BlocksRange {
+            start_block: current_hash,
+            end_block: block_hash,
```
Comment (rtsisyk) [inline node/src/chain_loader.rs]: This limit is never used in the code.

File: consensus/src/state.rs [hunk 5]
```diff
@@ -278,9 +278,6 @@ impl<Request: Hashable + Clone + Debug, Proof: Hashable + Clone + Debug> Consens
             return Err(ConsensusError::UnknownMessagePeer(msg.pkey));
         }
 
-        // Validate signature and content.
-        msg.validate()?;
```
Comment (rtsisyk) [inline consensus/src/state.rs]: Is it intentional?

File: node/src/lib.rs [hunk 6]
```diff
@@ -674,14 +674,45 @@ impl NodeService {
...
-
+    fn handle_sealed_block(&mut self, msg: SealedBlockMessage) -> Result<(), Error> {
         // Check signature and content.
         msg.validate()?;
 
+        let ref block = msg.block;
+        let block_hash = Hash::digest(block);
+
+        if let Block::MonetaryBlock(_) = block {
+            // TODO: Should this check exist? We can send block as response, and it would be validated without this check.
```
Comment (rtsisyk) [inline node/src/lib.rs]: Which one? `if let Block::MonetaryBlock(_) = block`?
Monetary block should be signed by the epoch leader. KeyBlock should be signed by a leader elected by VRF. It is a good question to check VRF leader.

File: node/src/lib.rs [hunk 7]
```diff
@@ -674,14 +674,45 @@ impl NodeService {
...
+                return Err(NodeError::SealedBlockFromNonLeader(
+                    block_hash,
+                    self.chain.leader.clone(),
+                    msg.pkey,
+                )
+                .into());
+            }
+        }
+
+        {
```
Comment (rtsisyk) [inline node/src/lib.rs]: Probably these brackets are not more needed in Rust 2018.

File: node/src/chain_loader.rs [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+            ChainLoaderMessage::Response(r) => {
+                "response".hash(state);
+                r.hash(state)
+            }
+        }
+    }
+}
+
+/// Limit of blocks to download starting from current known blockchain state.
+const ORPHANS_BLOCK_LIMIT: u64 = 1000;
```
Comment (rtsisyk) [inline node/src/chain_loader.rs]: Too many. Master can be easily DDoSed by such requests. Probably good value is 10 or 100..

File: node/src/chain_loader.rs [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+
+impl ChainLoader {
+    pub fn new() -> Self {
+        ChainLoader {
+            blocks_queue: Vec::new(),
+        }
+    }
+}
+
+//TODO: Download blocks from different participant.
```
Comment (rtsisyk) [inline node/src/chain_loader.rs]: Please don't forget to open tickets for this problems after finishing with this pull request.

File: blockchain/src/blockchain.rs [hunk 10]
```diff
@@ -245,6 +245,20 @@ impl Blockchain {
         return None;
     }
 
+    /// Returns blocks history starting from block_hash, limited by count.
+    pub fn blocks_range(&self, starting_hash: &Hash, count: u64) -> Option<&[Block]> {
```
[the remark this answers is not in the document]
Comment (vldm) [reply blockchain/src/blockchain.rs]: @rtsisyk Yes, but we already have blocks() method that return &[Block] and many code related to it len and context. We can rewrite this code to return Iterator, or something more specific. Do you want me to return iterator from rocksdb in this pullrequest, or we can deal with it after second version of persistance will be implemented?

File: node/src/chain_loader.rs (hunk 2 above)
Comment (vldm) [reply node/src/chain_loader.rs]: @rtsisyk I started with implementing more complex bootstrap system, in which this structure is used as part of responses too.
For now i can simplify it, or if we want to implement more complex bootstrap system in near future, we can keep this structure for future refactor?

File: consensus/src/state.rs (hunk 5 above)
Comment (vldm) [reply consensus/src/state.rs]: Yes, `feed_message` is internal method that handles internal (created by this node) and external messages. I refactored queue a bit, and found that validate could be called more then once for messages in queue. So i decide to move verify into `handle_consensus_message`

File: node/src/lib.rs (hunk 6 above)
Comment (vldm) [reply node/src/lib.rs]: @rtsisyk Okay, i get your point.

But this comment was about difference between Block and Sealed block message. We only keep signature of the leader in the sealed block message. But in the storage we only keep Block without leader signature, so when we sending blocks as BlocksResponse, we omit leader signature.

So do we really need to check leader signature? Its looks like redurant.

**Pull request merged**